Display a "purchase success" notice in Customer Home

niranjan-uma-shankar opened the pull request:
#### Changes proposed in this Pull Request

* When a new user lands in Customer Home after successfully purchasing a plan, show a success notice. Otherwise there's no indication that the transaction was successful.
* A minor UI fix that removes the space before "/month" when displaying price in checkout.

**New signup -> Checkout -> Customer Home**

<img width="1250" alt="Screenshot 2021-01-29 at 1 04 03 PM" src="https://user-images.githubusercontent.com/1269602/106245475-825dbc00-6232-11eb-833b-276453d86ee0.png">

**Launch flow -> Checkout -> Customer Home**

<img width="1235" alt="Screenshot 2021-01-29 at 12 45 55 PM" src="https://user-images.githubusercontent.com/1269602/106245519-92759b80-6232-11eb-8f33-7ad8865936a2.png">


#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Signup as a new user, purchase a plan, decline any upsell offers. Verify that you are taken to Customer Home with a success notice as shown in the screenshots above.
* Enter the site launch flow, purchase a plan, and verify that you are taken to Customer Home with a success notice.
* Signup as a new user, select a free plan, and confirm that you _do not see_ any notices when on Customer Home.
* Repeat step 1 with a non-EN language and verify that the success notice appears translated:
<img width="1264" alt="Screenshot 2021-01-29 at 12 58 21 PM" src="https://user-images.githubusercontent.com/1269602/106245375-5b06ef00-6232-11eb-890e-47f04a696624.png">
* Verify that in the checkout page, the price is displayed as "$4/month" where there's no space before 
"/month". Verify that in non-EN, translations are not affected.

<img width="571" alt="Screenshot 2021-02-03 at 3 59 45 PM" src="https://user-images.githubusercontent.com/1269602/106734321-e6202480-6638-11eb-8478-b93d35b795fe.png">

niranjan-uma-shankar committed:
Check if destination is eligible for signup destination

niranjan-uma-shankar committed:
1. Clear signup destination cookie when declining the upsell offer 2. Direct load of upsell offer and subsequent decline should take to root path /

niranjan-uma-shankar committed:
Revert

niranjan-uma-shankar committed:
Update test case

niranjan-uma-shankar committed:
Redirect to root url '/' if cart and receipt ID is empty

niranjan-uma-shankar committed:
Fix the case of clicking Decline on the offer page when opening with a cart item

niranjan-uma-shankar committed:
Show purchase success notice in Customer Home when arriving after completing a successful checkout purchase

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.js:
```diff
@@ -538,13 +538,13 @@ function LineItemSublabelAndPrice( { item } ) {
 
 	if ( item.type === 'plan' && item.wpcom_meta?.months_per_bill_period > 1 ) {
-		return translate( '%(sublabel)s: %(monthlyPrice)s /month × %(monthsPerBillPeriod)s', {
+		return translate( '%(sublabel)s: %(monthlyPrice)s/month × %(monthsPerBillPeriod)s', {
```
:ok: This change will be queued for retranslation. We'll use the existing translations in the meantime.

niranjan-uma-shankar commented:
> I think that it's actually more common in English to not have a space before and after the / (similar example). Could you remove both and add a comment?

> As already mentioned, since this function is 100% covered by tests, it would be great if you could add a test to cover this new behavior!

@michaeldcain @sirbrillig I've addressed your review comments posted in https://github.com/Automattic/wp-calypso/pull/49612 in this PR.

sirbrillig commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -56,6 +58,13 @@ const Home = ( {
 
+	if ( 'purchase-success' === noticeType ) {
+		const successMessage = translate( 'Your purchase has been completed!' );
+		notices.success( successMessage, {
```
There's currently a project to Reduxify the `notices` module (see https://github.com/Automattic/wp-calypso/issues/48408) (cc @tyxla) so we should probably do this with `const reduxDispatch = useDispatch();` and `reduxDispatch( successNotice( /* ... */ ) )` instead. Be aware that the `persistent` option has become `isPersistent`, which tripped me up before.

tyxla replied:
Thanks for addressing this early @sirbrillig @niranjan-uma-shankar, the old notices module is almost gone now (and deprecated as per [its README](https://github.com/Automattic/wp-calypso/tree/trunk/client/notices#notices)) so Ideally we shouldn't introduce new instances of it.  👍

niranjan-uma-shankar committed:
Reduxify notice

niranjan-uma-shankar committed:
Fix merge error

sirbrillig commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -56,6 +58,16 @@ const Home = ( {
 	}
 
+	if ( 'purchase-success' === noticeType ) {
+		const reduxDispatch = useDispatch();
```
This probably will throw an error because React hooks cannot be used inside a condition. This is partially my fault for not being clear in my suggestion. If you move this line up to line 50 then it should work.

niranjan-uma-shankar replied:
I've added a new commit that fixes this 👍

niranjan-uma-shankar committed:
Move useDispatch outside of conditional

sirbrillig commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.js:
```diff
@@ -538,13 +538,13 @@ function LineItemSublabelAndPrice( { item } ) {
 				monthsPerBillPeriod: item.wpcom_meta.months_per_bill_period,
 			},
-			comment: 'product type and monthly breakdown of total cost, separated by a colon',
+			comment: 'product type and monthly breakdown of total cost, separated by a colon. For "/month", there is no space before and after the "/" but please add a space if it makes sense in other languages.',
```
Very minor: the Prettier eslint tests are failing on this line because of something to do with whitespace. If you run Prettier on this file (should happen automatically when committing) it should fix it.

niranjan-uma-shankar replied:
Done in [9ee4b5f](https://github.com/Automattic/wp-calypso/pull/49612/commits/9ee4b5ff00721429b313921b465b0674246eb076).

sirbrillig commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -56,6 +59,15 @@ const Home = ( {
 
+	if ( 'purchase-success' === noticeType && layout ) {
+		const successMessage = translate( 'Your purchase has been completed!' );
+		reduxDispatch(
```
Hm... one more thing I was thinking of. This might be fine but I wonder if this should happen on every render, or only when `noticeType` changes? I'm thinking that if it happens on every render as it is currently that it might recreate the notice even if it's dismissed. I'm not sure.

Anyway, if you'd like to do this only when `noticeType` changes, something like this should work:

```js
useEffect(() => {
  if (! layout || noticeType !== 'purchase-success') {
    return;
  }
  const successMessage = translate( /* ... */ );
  reduxDispatch( successNotice( /* ... */ ) );
}, [noticeType, layout, reduxDispatch, translate]);
```

niranjan-uma-shankar replied:
Good point! I've added a fix for this in [6bccc79](https://github.com/Automattic/wp-calypso/pull/49612/commits/6bccc794ba504cc70032d9d52eec60625cc231fa).

niranjan-uma-shankar committed:
Render notice only if the notice type has changed

sirbrillig commented on client/my-sites/customer-home/main.jsx:
```diff
@@ -49,6 +49,18 @@ const Home = ( {
+				isPersistent: true,
+			} )
+		);
+	}, [ noticeType, layout, reduxDispatch, translate ] );
```
Minor: I think `canUserUseCustomerHome` needs to be in the dependencies array for this (possibly helpful: if you set up your editor to show eslint warnings, this should show up) since that's used in the block. It _probably_ won't matter (I imagine that variable won't change without other variables changing too) but it's nice not to have extra warnings in the code.

gwwar replied:
So another way of approaching this is that we can dispatch a global notice as a side-effect. This can be done from this specific checkout data handling, or as a side effect of the [notice](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/notices/middleware.js) or [navigation](https://github.com/Automattic/wp-calypso/blob/trunk/client/state/navigation/middleware.js) middleware. (Generally. I'd try the former first, especially if the action isn't general enough to warrant middleware handling.)

If you're interested, let me know if you need a working example. @dmsnell is also usually available for advice on data-layer handling.

tyxla replied:
FWIW we're currently in the process of removing the notice middleware completely. I'd advise not introducing new notices that are implemented with it. Let me know if you need additional details or rationale.

gwwar replied:
Thanks @tyxla Are we still supporting more complex side effects in the data-layer folder https://github.com/Automattic/wp-calypso/tree/trunk/client/state/data-layer/wpcom

That's what I'd try first for showing a success notice after a purchase.

tyxla replied:
@gwwar Sure, dispatching a notice that corresponds to an action type in the data layer is just fine.  👍 

The primary reason for us to want to remove the notices middleware is that it prevents state from being fully modularized. Data layer, on the other hand, is modular already, so as long as we don't rely on that functionality at boot time, we're good.

gwwar replied:
Thanks @tyxla, and 🚀 for the data modularization efforts!

sirbrillig reviewed (approved):
This looks good to me code-wise (with one minor comment)! I'll leave running the testing instructions to those on this PR more familiar with this area.

gwwar commented:
General question: Why would we not show a purchase success notice if the site type doesn't enable My Home? For example on Calypso Green after a plan purchase, is there no confirmation?

niranjan-uma-shankar committed:
Remove EOF newline

niranjan-uma-shankar commented:
> Why would we not show a purchase success notice if the site type doesn't enable My Home? For example on Calypso Green after a plan purchase, is there no confirmation?

Good point. The reasoning behind keeping the notice specific to "My Home" is that in other places such a notice could appear repetitive. Take the following cases:
* On an existing free plan account, if I upgrade to a paid plan, I am taken to a "Thank you" page that confirms my purchase. 
* If I signup as a new user and purchase an eCommerce plan, then after the transaction completes, I'm taken to a page that asks to wait (while atomic transfer happens in the background) and then I'm taken to an [Atomic Thank You page](https://d.pr/i/BoQqLk) that confirms my successful purchase.

The same is true for Jetpack - we take the user to a Jetpack-specific page that describes next steps for their purchase. 

How do you feel about this @gwwar?

niranjan-uma-shankar committed:
Add canUserUseCustomerHome to the dependencies array

gwwar commented:
> The same is true for Jetpack - we take the user to a Jetpack-specific page that describes next steps for their purchase.

That's reasonable, I'd still recommend dropping the need to pass through the notice as a prop to main files. I'd check to see if we can dispatch a global notice in a wpcom data-layer success handler. See examples in: https://github.com/Automattic/wp-calypso/tree/trunk/client/state/data-layer/wpcom

What endpoint do we post to for checkout? It's probably possible to drop the cookie reading as well.

gwwar commented:
If this is urgent, these changes look okay, but I'd encourage a follow up PR as it may be good opportunity to learn a bit more about handling options in the data-layer.

niranjan-uma-shankar commented:
> Why would we not show a purchase success notice if the site type doesn't enable My Home?

I'll also add another scenario here. After completing a plan purchase in the signup flow, we show an upsell offer to most users(either for Quick Start or for upgrading to the next higher plan). I believe we wouldn't want to show a purchase success notice on the offer page since we want to hold the attention of the user a wee bit longer as if the payment is still processing in the background. We would want to show the success notice for their plan purchase if they decline the upsell offer and land in Customer Home. For this reason, I'm not sure if triggering the notice in a checkout endpoint success callback would work in all scenarios. @victorespigares I would be interested in your take here. Would a "Your purchase has been completed!" notice of the type shown in the image in the PR description be appropriate on an upsell offer page when the user is coming from a plan purchase?

niranjan-uma-shankar committed:
Force prettier

gwwar commented:
> How do you feel about this @gwwar?

My notes aren't blocking. If y'all have thought through use cases, feel free to move forward 👍

victorespigares commented:
> I believe we wouldn't want to show a purchase success notice on the offer page since we want to hold the attention of the user a wee bit longer as if the payment is still processing in the background. We would want to show the success notice for their plan purchase if they decline the upsell offer and land in Customer Home.

This is consistent with how we've been approaching upsell offers so far. They happen in the interim stage between the user pressing "Pay" and the confirmation that's coming in the next step.

> @victorespigares I would be interested in your take here. Would a "Your purchase has been completed!" notice of the type shown in the image in the PR description be appropriate on an upsell offer page when the user is coming from a plan purchase?

We haven't really tested it, so it's hard to tell. The closest to this is what's going to be launch in the DIFM upsell offer in https://github.com/Automattic/wp-calypso/pull/49301/. That's the first post-purchase upsell offer that includes a purchase confirmation notice within the offer, in that case, because it improved the flow since it doesn't lead into Customer Home. Unfortunately, it's a very different type of offer (lead-gen, doesn't take to Customer Home, etc.) so it won't be apples to apples. Ideally, this is a change we should test down the line with QS and plan bumps.

niranjan-uma-shankar committed:
Fix merge conflict

a8ci18n commented on client/my-sites/checkout/composite-checkout/components/wp-order-review-line-items.tsx:
```diff
@@ -570,13 +570,14 @@ function LineItemSublabelAndPrice( {
 		return (
 			<>
-				{ translate( '%(sublabel)s: %(monthlyPrice)s /month × %(monthsPerBillPeriod)s', {
+				{ translate( '%(sublabel)s: %(monthlyPrice)s/month × %(monthsPerBillPeriod)s', {
```
:ok: This change will be queued for retranslation. We'll use the existing translations in the meantime.
